### ISSUE TITLE
Output values being referenceable

### ISSUE THREAD

**Issue by ericdmoore** [federalies/squals#4]
Often you need to make an S3 Bucket - reference that implied domain name as an `origin` for CDN - and then repeatedly - setup the domain from the CDN into a Route53 A or CNAME record... this is hard  to do right now...

Think through the architecture so that this very normal use case becomes obvious...

Maybe by nesting during declaration in a recipe or component show the hierarchy which would then be flattened when serialized into the template...

**Comment by ericdmoore** [comment]
What if it was like

```javascript
route53.CNAME({
  ref: cloudFrontCDN.origins[
      S3Bucket()
  ]
})
```

**Comment by ericdmoore** [comment]
The real issue is there is a conceptual [DAG](https://en.wikipedia.org/wiki/Directed_acyclic_graph) running through the resources (or modeled objects) which are only partially defined at the time of declaration - and they become full resources at run-time.

**Comment by ericdmoore** [comment]
Why is this important? 
AWS-CDK is trying to solve this too because to ship software as a declared stack is pretty enticing with clean borders of how to "install" packages (less in a language sense of the word - "package") but more in most broad sense of a collection of resources that accomplish a purpose.

**Comment by ericdmoore** [comment]
Or, for example maybe S3Bucket() would return an object with properties + attributes in the `cloudformation` sense of those words - where attributes return a function - to be called later? or perhaps its a Promise resolved ...later?

**Comment by ericdmoore** [comment]
So far it is easiest to add getters to each class - which does keep it more imperative.
so its more like:

```javascript
const s = new S3Bucket()
console.log(s.DomainName()) 
```

and actually since the domain does not exist until template is deployed... the s.DomainName() is actually a helper function assuming the property is being referenced...

so it returns `{ 'Fn::GetAtt': [this.logicalName, 'Arn'] }`

**Comment by ericdmoore** [comment]
This is roughly addressed at least by commit:37ea11fc89daa4d683f529c6405874b97b9f9c41

but there is a greater question of how to create a functional pattern matching on the various types you might want to generate a new template from... (opening a new issue to track that one)
